execution.go

```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
-	rollkitStateRoot := execution_types.Hash(stateRoot[:])
-
-	gasLimit := payloadResult.ExecutionPayload.GasLimit
-
-	c.initialHeight = initialHeight
-
-	return rollkitStateRoot, gasLimit, nil
+	return stateRoot[:], gasLimit, nil
 }
 
-// GetTxs retrieves transactions from the transaction pool
-func (c *EngineAPIExecutionClient) GetTxs(ctx context.Context) ([]execution_types.Tx, error) {
-	var result struct {
-		Pending map[string]map[string]*types.Transaction `json:"pending"`
-		Queued  map[string]map[string]*types.Transaction `json:"queued"`
-	}
-	err := c.ethClient.Client().CallContext(ctx, &result, "txpool_content")
+// GetTxs retrieves transactions from the current execution payload
+func (c *PureEngineClient) GetTxs(ctx context.Context) ([][]byte, error) {
+	var payloadResult engine.ExecutionPayloadEnvelope
+	err := c.engineClient.CallContext(ctx, &payloadResult, "engine_getPayloadV3", c.payloadID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get tx pool content: %w", err)
+		return nil, fmt.Errorf("engine_getPayloadV3 failed: %w", err)
 	}
 
-	var txs []execution_types.Tx
+	// Store the original transactions
+	originalTxs := payloadResult.ExecutionPayload.Transactions
 
-	// add pending txs
-	for _, accountTxs := range result.Pending {
-		for _, tx := range accountTxs {
-			txBytes, err := tx.MarshalBinary()
-			if err != nil {
-				return nil, fmt.Errorf("failed to marshal transaction: %w", err)
-			}
-			txs = append(txs, execution_types.Tx(txBytes))
-		}
+	// Clear transactions before serializing
+	payloadResult.ExecutionPayload.Transactions = [][]byte{}
+
+	jsonPayloadResult, err := json.Marshal(payloadResult)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize payloadResult: %w", err)
 	}
 
-	// add queued txs
-	for _, accountTxs := range result.Queued {
-		for _, tx := range accountTxs {
-			txBytes, err := tx.MarshalBinary()
-			if err != nil {
-				return nil, fmt.Errorf("failed to marshal transaction: %w", err)
-			}
-			txs = append(txs, execution_types.Tx(txBytes))
-		}
+	// Create the result with serialized payload as first tx, followed by original transactions
+	txs := make([][]byte, len(originalTxs)+1)
+	txs[0] = jsonPayloadResult
+	for i, tx := range originalTxs {
+		txs[i+1] = tx
 	}
+
 	return txs, nil
 }
 
-func (c *EngineAPIExecutionClient) getBlockHash(ctx context.Context, height uint64) (common.Hash, error) {
-	var block map[string]interface{}
-	err := c.engineClient.CallContext(
-		ctx, &block, "eth_getBlockByNumber", rpc.BlockNumber(height), false) //nolint:gosec // disable G115
+// ExecuteTxs executes the given transactions at the specified block height and timestamp
+func (c *PureEngineClient) ExecuteTxs(ctx context.Context, txs [][]byte, blockHeight uint64, timestamp time.Time, prevStateRoot []byte) (updatedStateRoot []byte, maxBytes uint64, err error) {
+	var payloadResult engine.ExecutionPayloadEnvelope
+	// First tx is the serialized payload
+	firstTx := txs[0]
```

Comment:
nit: worth validating the len of txs > 0 to avoid Panics due to developer errors.

Reply:
Follow-up in #35